specs/consensus/overview.md

```diff
@@ -437,18 +437,134 @@ height of consensus `h_p`.
+different semantics of different blocks based on the decision round. We strongly
+suggest to not use round numbers in validation of values for this reason.
+
+#### Backward compatibility of `valid(v)`
+
+**Requirement 1 (Fixing bugs).**
+There might be a bug in `valid`. Then it might be possible that due to a bug,
+`valid` returns `false` 
+for a value `v` proposed by a correct process, and we are stuck at a given height. 
+A way to get out of the 
+situation is to produce a new implementation `valid'` that returns `true` for `v`. So
+to be prepared for such a scenario we need to allow a change in the function.
+
+**Requirement 2 (Future use).**
+So if we allow changes to `valid`, we need to understand all uses of this function: Some
+synchronization protocols may use `valid` for consistency checks, for instance, if a node
+fell behind, it might need to learn several past decisions. In doing so, it typically also
+uses (the current version) of `valid` to check the decided values before accepting them.
+In this scenario, a value decided in the past (potentially using a now old version of 
+`valid`) should be deemed valid with the current version of the function. 
+
+These two requirements lead us to the following requirement on the implementations:
+we consider a sequence of `valid_i(v)` implementations, with increasing indexes `i`, 
+so that to represent multiple _backwards compatible_ implementations of the validity checks.
+Formally we require that
+
+`valid_i(v) == true` implies `valid_j(v) == true `, for `j > i` and every value `v`
+
+The logical implications allows to be more permissive in later versions (as required to
+fix the bug), while maintaining the need that newer versions allow to check validity
+of previously decided values.
+
+#### Backward compatibility of `valid(v, chain, height)`
+
+As mentioned above, a typical use case is that the validity depends on the 
+state of the blockchain at a given height. In this scenario, the two requirements
+from above can be achieved without the implication:
+The idea is that if `valid` should be changed in that from height 
+`h`on, a function `new_valid` should be used to check validity, while up to height
+`h-1` the function `old_valid` has been used, then we can get something 
+along these lines:
+```
+valid(v, chain, height) = 
+  if (height >= h)
+    new_valid(v, chain, height)
+  else
+    old_valid(v, chain, height)
+```
+
+In this way, we effectively split evaluation of `valid` into different height 
+intervals. Whenever
+a new version is needed, and new interval is added on top. Moreover
+if in
+the future validity of a value is checked with a new version of `valid`, effectively
+the logic that was in place for that height is used by using the proper "if" branch. 
+This addresses Requirement 1 and 2 from above.
+
+TODO: I think in this [remark](https://github.com/informalsystems/malachite/issues/510#issuecomment-2505632410), different use cases (soft upgrades) are mixed.
```

Comment:
Indeed, the example in the code above is almost the definition of software upgrades.

Reply:
The linked comment is a short-back of the code snippet and the software upgrades approach. This is the case where a bug identified during a given height forces us to the most general approach described in the "Backwards Compatibility" section.

Reply:
I updated the discussion accordingly.